fix(gatsby): granular redux cache heuristics for finding large nodes

pvdz opened the pull request:
When persisting the redux store, there is a heuristic that will try to roughly determine how big nodes are because there are intrinsic limits to the amount of data we can store in one pass (2gb). This heuristic is currently taking 11 random nodes from the entire pool and using the biggest one to determine the chunk size.

This PR improves the granularity of this check to do this per node type, because certain nodes are intrinsically smaller than others so it hopefully yields more relevant information. I suspect that, ultimately, we could skip all of them and just check the Page type, but we can determine that later.

One downside is that this means sites with many types, like Contentful sites, will probe 11*x nodes now. I don't expect this to be a huge problem, but something noteworthy regardless.

Fixes #23627

This fix will not work with Loki, since that doesn't create the by-type map. Since I'm planning to remove that in the next two weeks I'm probably going to wait with actually getting this PR merged until that happens.

freiksenet commented:
> we could skip all of them and just check the Page type

I suspect that too, tbh. Generally the plan is to start warning about big Page types because it usually means big context and that is bad for runtime performance.

pvdz commented:
I fixed a bug, added a test. I think we can't merge this because of the perf implication :'(

The problem is that the map by type does not have arrays but it has another map. This means we can't have direct access to the nodes and we either need to create a temporary array for this, or loop through all nodes manually, either way a performance penalty.

I think we'll need to find another way.

pvdz commented:
Just not worth it currently so I'm going to close this.
